### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by FlyingBird95**
Major refactoring of the Dashboard.

Most important changes:

- The Jinja2-templates now uses only resources for JQuery and Bootstrap. This reduces the project with 50k JS lines. Bootstrap is updated from 3.0 to 4.0. Therefore, the layout is slightly changed.
- Removed two graphs: hits per hour and execution time per hour:
- On the overview page, the average is replaced with the median execution time.
- Added two types of forms to reduce the graph:
   - A graph that shows data across a number of days can be reduced by a 'Date-range-form', that is: a start-date and an end-date.
   - A graph that plots a number of items (e.g. the Time per version per User) can be reduced by selecting an integer that represent the amount of users that are displayed.
- Outlier page is divided into subpages, each page shows at most 10 outliers.
- Group by is refactored, it now accepts any combination of: lists, int, strings, tuples and functions.

Resolved issues:
- https://github.com/flask-dashboard/Flask-MonitoringDashboard/issues/63
- https://github.com/flask-dashboard/Flask-MonitoringDashboard/issues/77
- https://github.com/flask-dashboard/Flask-MonitoringDashboard/issues/80
- https://github.com/flask-dashboard/Flask-MonitoringDashboard/issues/82
- https://github.com/flask-dashboard/Flask-MonitoringDashboard/issues/89

### CODE REVIEW

**Comment by bogdanp05** [comment]
@mircealungu I'll be able to do it tomorrow.

**Comment by mircealungu** [comment]
Thanks @bogdanp05 ! this is cool!

File: flask_monitoringdashboard/database/count.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+def count_requests(endpoint, *where):
+    """ Return the number of hits for a specific endpoint (possible with more filter arguments).
+    :param endpoint: name of the endpoint
+    :param where: additional arguments
+    """
+    return count_rows(FunctionCall.id, FunctionCall.endpoint == endpoint, *where)
+
+
+def count_outliers(endpoint):
```
Comment (bogdanp05) [inline flask_monitoringdashboard/database/count.py]: Couldn't some of these functions be concatenated into a single function like `count(endpoint, what_to_count)`, where `what_to_count` would be users, outliers, etc.?

File: flask_monitoringdashboard/database/function_calls.py [hunk 2]
```diff
@@ -4,73 +4,50 @@
...
-            filter(FunctionCall.time > date_from).group_by(FunctionCall.endpoint).all()
-        db_session.expunge_all()
-        return result
+        db_session.add(FunctionCall(endpoint=endpoint, execution_time=execution_time, version=config.version, ip=ip))
 
 
-def get_average(date_from=None):
-    """ Return the average of execution times which are called after 'date_from' grouped per endpoint
-    :param date_from: A datetime-object
+def get_median(endpoint, *where):
```
Comment (bogdanp05) [inline flask_monitoringdashboard/database/function_calls.py]: This function makes 2 queries to the db: first on line 44 and the second on line 46.

File: flask_monitoringdashboard/database/function_calls.py [hunk 3]
```diff
@@ -4,73 +4,50 @@
...
-        result = db_session.query(FunctionCall.endpoint,
-                                  func.avg(FunctionCall.execution_time).label('average')). \
-            filter(FunctionCall.time > date_from).group_by(FunctionCall.endpoint).all()
-        db_session.expunge_all()
-        return result
+        result = db_session.query(FunctionCall.execution_time). \
+            filter(FunctionCall.endpoint == endpoint, *where). \
+            order_by(FunctionCall.execution_time).limit(2 - num_rows % 2).offset((num_rows - 1) / 2).all()
+        values = [r[0] for r in result]
+        return sum(values) / max(1, len(values))
```
Comment (bogdanp05) [inline flask_monitoringdashboard/database/function_calls.py]: no `db_session.expunge_all()` ?

File: flask_monitoringdashboard/database/outlier.py [hunk 4]
```diff
@@ -15,9 +15,25 @@ def add_outlier(endpoint, execution_time, stack_info, request):
...
+    """
     with session_scope() as db_session:
-        result = db_session.query(Outlier).filter(Outlier.endpoint == endpoint).order_by(desc(sort_column)).all()
+        result = db_session.query(Outlier).filter(Outlier.endpoint == endpoint).order_by(desc(sort_column)).\
+            offset(offset).limit(per_page).all()
         db_session.expunge_all()
         return result
+
+
+def delete_outliers_without_stacktrace():
```
Comment (bogdanp05) [inline flask_monitoringdashboard/database/outlier.py]: Why delete them? The still provide CPU and memory info so they might be useful.

File: flask_monitoringdashboard/database/function_calls.py [hunk 5]
```diff
@@ -4,73 +4,50 @@
...
-    with session_scope() as db_session:
-        result = db_session.query(FunctionCall.endpoint,
-                                  func.count(FunctionCall.execution_time).label('count')). \
-            filter(FunctionCall.time > date_from).group_by(FunctionCall.endpoint).all()
-        db_session.expunge_all()
-        return result
+        db_session.add(FunctionCall(endpoint=endpoint, execution_time=execution_time, version=config.version, ip=ip))
 
 
-def get_average(date_from=None):
```
Comment (bogdanp05) [inline flask_monitoringdashboard/database/function_calls.py]: Why not also keep the `get_average` function? #77 says the plot should include both median and average.

File: flask_monitoringdashboard/core/utils.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+
+def simplify(values, n=5):
+    """
+    Simplify a list of values. It returns a list that is representative for the input
+    :param values: list of values
+    :param n: length of the returned list
+    :return: list with n values: min, q1, median, q3, max
+    """
+    return [np.percentile(values, i * 100 // (n-1)) for i in range(n)]
```
Comment (bogdanp05) [inline flask_monitoringdashboard/core/utils.py]: Can n be 1?

**Comment by mircealungu** [comment]
thanks @bogdanp05 for the review. 
you should have let @FlyingBird95 to address your feedback before merging though :) 
in any case he can still address the problems that need to be addressed later. 
in the meantime, i've deployed the new version, and I like it. congrats for the nice work @FlyingBird95 !!

**Pull request merged**